### PR TITLE
[ALLUXIO-3316][WIP] Replace FSMaster thrift service w/ gRPC

### PR DESCRIPTION
**Pull request by madanadit**
https://alluxio.atlassian.net/browse/ALLUXIO-3316

### CODE REVIEW

File: core/transport/src/main/java/alluxio/util/grpc/GrpcChannel.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+package alluxio.util.grpc;
+
+import io.grpc.CallOptions;
+import io.grpc.Channel;
+import io.grpc.ClientCall;
+import io.grpc.MethodDescriptor;
+
+/**
+ * A simple wrapper around the {@link Channel} class in grpc. Outside of this module, this
+ * class should be used to replace references to {@link Channel} for dependency management.
```
Comment (aaudiber) [inline core/transport/src/main/java/alluxio/util/grpc/GrpcChannel.java]: why is this useful for dependency management?

**Comment by madanadit** [comment]
Thanks @aaudiber for taking a look.
- How about `alluxio-core-api` instead of `alluxio-core-base`? This module was split from transport because to remove a cyclic dependency b/w transport and common. The general guideline for what goes in base is the service interfaces and any dependencies
- Makes sense. Will move to proto3 in a follow up PR
- The `FileSystemMasterOptions` interface is used by the transport module to instantiate options with the server side defaults (which is done in the master module)

File: core/transport/src/main/java/alluxio/util/grpc/GrpcChannel.java (hunk 1 above)
Comment (madanadit) [reply core/transport/src/main/java/alluxio/util/grpc/GrpcChannel.java]: the invariant is that there should be no references to `io.grpc` outside the transport module and we relocate `io.grpc` when building transport

**Comment by aaudiber** [comment]
- +1 for `alluxio-core-api`. I think right now there's a lot of stuff in there that belongs back in `core-common`, e.g. `IndexedSet`.

I don't like how we have master-specific logic in the `core-transport` module. What we want is to pass partial options over the wire, then have the master fill in the rest using its own defaults. But we can't do it that way currently because we don't have a concept of partial options, except in the transport layer. So the current code works around that by having options go from master to transport, then back to master. I see a few ways to get around this:

1. Pass proto options (partial options) to the master API, and let it resolve them against its defaults, then pass full options to the rpc implementations. This requires shading proto within the master module.
2. Get rid of our non-proto options completely and use proto. This saves us a ton of hand-written boilerplate, and lets us take advantage of the proto object APIs which are much more comprehensive than our hand-written options. We can implement client-side defaults by updating unset fields after the user passes us their options. This approach will require shading proto in more places.
3. Update our options to act as partial options (e.g. box all fields in `Optional`), then pass those from `transport` to `master`.
4. Add a partial options corresponding to all of our complete options. This doubles our boilerplate, but contains `proto` to `transport`.

I prefer (2) because it minimizes duplication and boilerplate, and makes options easier to work with. (1) has the advantage of letting us differentiate between partial and complete options, but at the cost of a bunch of boilerplate.

Basically, we can pay one layer of boilerplate to avoid the dependency on proto (what we are currently doing), and we can pay another layer to differentiate between partial and complete options (not currently doing this, which is why we got into the `master` -> `transport` -> `master` pretzel).

File: core/transport/src/main/java/alluxio/util/grpc/GrpcChannel.java (hunk 1 above)
Comment (aaudiber) [reply core/transport/src/main/java/alluxio/util/grpc/GrpcChannel.java]: ah, that makes sense now. Do you know why we we chose to shade/relocate within only one project instead of across the uber-jar? It seems like this way has more boilerplate like this, and is also more error-prone since we can accidentally reference `io.grep` from outside this module by accident

Comment (madanadit) [reply core/transport/src/main/java/alluxio/util/grpc/GrpcChannel.java]: we're fine on the server side but the client side side was messy and had issues (don't remember exactly what messed up though). 

we had the same kind of issue with proto, and the solution of wrapping grpc classes is an extension of what we did there. in the present code, there are no proto references outside the protobuf module (now renamed to transport)

**Comment by madanadit** [comment]
yeah there's a lot more in `alluxio-core-api` that doesn't belong there. will remove stuff in follow up PRs, the refactor was becoming a little too bug
